### PR TITLE
sql: support COMMENT ON SCHEMA

### LINKED ISSUE #67689: sql: Support 'COMMENT ON SCHEMA'
**Is your feature request related to a problem? Please describe.**
`COMMENT ON SCHEMA` is not supported.

**Describe the solution you'd like**
`COMMENT ON SCHEMA` is supported.

Support for `COMMENT ON` for tables, table columns, and databases was implemented in 2019 for https://github.com/cockroachdb/cockroach/issues/19472 then the issue was closed by @jordanlewis because schemas were not supported at all: https://github.com/cockroachdb/cockroach/issues/19472#issuecomment-486087994
> This is done but for schemas which we don't support anyway - closing. 

`CREATE SCHEMA` was implemented in 2020 https://github.com/cockroachdb/cockroach/pull/50613 therefore `COMMENT ON SCHEMA` is (should be?) unblocked. Might be a good first PR as there are examples in:
* [sql: support COMMENT ON DATABASE](https://github.com/cockroachdb/cockroach/pull/33057)
* [sql: support COMMENT ON COLUMN](https://github.com/cockroachdb/cockroach/pull/33355)

*I may implement the PR, but am not claiming it yet.*

**Describe alternatives you've considered**
I am skipping unsupported SQL.

### PR DESCRIPTION
**Pull request by ekalinin**
Fixes #67689 

This change adds support for SCHEMA comment.

Release note (sql change): This change adds associating
comment to SQL schema using PostgreSQL's 
`COMMENT ON SCHEMA` syntax.

### CODE REVIEW

**Review by knz — COMMENTED** [review]
This PR is missing the necessary changes to `sql/pg_catalog.go` to generate the values in `pg_catalog.pg_description`. This will also need a test about how to retrieve the schema comment via `pg_catalog.obj_description()`.

You'll probably need input from someone in the SQL schema team about how to populate pg_description properly.

> Reviewed 14 of 14 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-:-MgoylVr6jysV-lGfQu6:bdlorgr)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ekalinin)

___
*[pkg/sql/drop_schema.go, line 185 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-Mgox_9o74TAAlnFXGFe:-Mgox_9o74TAAlnFXGFf:b-8wbta1) ([raw file](https://github.com/cockroachdb/cockroach/blob/9008f0807135eb328f811a6628b907cd12821f2e/pkg/sql/drop_schema.go#L185)):*
> ```Go
> 		}
> 
> 		if err := p.removeSchemaComment(params.ctx, sc.GetID()); err != nil {
> ```

This needs to go to `dropSchemaImpl` otherwise the comment is not dropped upon DROP DATABASE CASCADE

**Comment by ekalinin** [comment]
> This PR is missing the necessary changes to `sql/pg_catalog.go` to generate the values in `pg_catalog.pg_description`. This will also need a test about how to retrieve the schema comment via `pg_catalog.obj_description()`.
> 
> You'll probably need input from someone in the SQL schema team about how to populate pg_description properly.
>

Thanks. Added.

> This needs to go to dropSchemaImpl otherwise the comment is not dropped upon DROP DATABASE CASCADE

Done.

**Review by knz — COMMENTED** [review]
> Reviewed 6 of 6 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ekalinin)

___
*[pkg/sql/pg_catalog.go, line 1398 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqkT9i2JaHPUeTuiSL:-MgqkT9i2JaHPUeTuiSM:b-mmvxyh) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/pg_catalog.go#L1398)):*
> ```Go
> 			case keys.SchemaCommentType:
> 				objID = tree.NewDOid(tree.MustBeDInt(objID))
> 				classOid = tree.NewDOid(catconstants.PgCatalogSchemaID)
> ```

We'll need a review from SQL Schema for this change.
(@ajwerner?)
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5085 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqkgGH-Rwkp3gtO3wW:-MgqkgGH-Rwkp3gtO3wX:b-vdsegf) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5085)):*
> ```
> ----
> objoid  classoid    objsubid  description
> 94      4294967022  0         testing schema
> ```

I think we need to test this via `pg_catalog.obj_description()` because I don't think we wish to assert particular object IDs inside unit tests.

Someone from SQL experience would know better.
(@rafiss?)

File: pkg/sql/sem/builtins/pg_builtins.go [hunk 1]
```diff
@@ -1942,6 +1942,8 @@ func getCatalogOidForComments(catalogName string) (id int, ok bool) {
 		return catconstants.PgCatalogClassTableID, true
 	case "pg_database":
 		return catconstants.PgCatalogDatabaseTableID, true
+	case "pg_schema":
+		return catconstants.PgCatalogSchemaID, true
```
Comment (postamar) [inline pkg/sql/sem/builtins/pg_builtins.go:1946]: I checked the postgres documentation and I couldn't find any mention of `pg_schema` in there. Strictly speaking the table for schemas in the pg catalog is `pg_namespace`. As far as comments are concerned, it seems they might actually belong in `pg_shdescription` or `pg_description`. Perhaps @rafiss can enlighten us?

Comment (knz) [reply pkg/sql/sem/builtins/pg_builtins.go:1946]: a quick test inside `psql` tells me that schema comments get stored inside `pg_description`.

**Comment by ekalinin** [comment]
> I think we need to test this via pg_catalog.obj_description() because I don't think we wish to assert particular object IDs inside unit tests.

Fixed.

> I checked the postgres documentation and I couldn't find any mention of pg_schema in there. Strictly speaking the table for schemas in the pg catalog is pg_namespace

Fixed.

**Review by otan — COMMENTED** [review]
> Reviewed 13 of 14 files at r1, 2 of 6 files at r2, 3 of 4 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ekalinin, @knz, and @rafiss)

___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5085 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqkgGH-Rwkp3gtO3wW:-MhGus0VC1LHoXxfRZty:b282zv1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5085)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

I think we need to test this via `pg_catalog.obj_description()` because I don't think we wish to assert particular object IDs inside unit tests.

Someone from SQL experience would know better.
(@rafiss?)
</blockquote></details>

i'd probably omit the oids / subids simply because they are potentially unstable.
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5082 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MhGvFV98nv8A46mYcnL:-MhGvFV98nv8A46mYcnM:b-um91z7) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b4e6fd8283f321fc119346558ad4644c6e7c14/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5082)):*
> ```
> SELECT obj_description(objoid)
>   FROM pg_catalog.pg_description
>  WHERE description = 'testing schema';
> ```

nit: drop the trailing `;` here and above
___
*[pkg/sql/logictest/testdata/logic_test/schema, line 881 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MhGvI9B97OtFUBj-9oI:-MhGvI9B97OtFUBj-9oJ:b2f5ohj) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b4e6fd8283f321fc119346558ad4644c6e7c14/pkg/sql/logictest/testdata/logic_test/schema#L881)):*
> ```
> 
> statement ok;
> USE comment_db;
> ```

nit: drop the trailing `;` here 
___
*[pkg/sql/sem/builtins/pg_builtins.go, line 1946 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqaSox7ndfLf79FuDJ-r2-1946:-MhGvUzB5Vr6PQ-l7aXI:b-ppk6vx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/sem/builtins/pg_builtins.go#L1946)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

a quick test inside `psql` tells me that schema comments get stored inside `pg_description`.
</blockquote></details>

`pg_description` is correct

**Review by ekalinin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ekalinin, @knz, and @rafiss)

___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5085 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqkgGH-Rwkp3gtO3wW:-MgvhatA2zWy-YJTmHbP:b-qoxx8q) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5085)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

i'd probably omit the oids / subids simply because they are potentially unstable.
</blockquote></details>

Should I remove test via `obj_description`?

**Comment by otan** [comment]
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5085 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqkgGH-Rwkp3gtO3wW:-MhISs1n0Oaj7J_D_umI:bqv3vp1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5085)):*
<details><summary><i>Previously, ekalinin (Eugene Kalinin) wrote…</i></summary><blockquote>

Should I remove test via `obj_description`?
</blockquote></details>

Is there a test for it? We should keep it for sure if there is and add it if it isn't.

**Review by ekalinin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ekalinin, @knz, @otan, and @rafiss)

___
*[pkg/sql/drop_schema.go, line 185 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-Mgox_9o74TAAlnFXGFe:-MhKBmKG9ZmdDc3JwuFl:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/9008f0807135eb328f811a6628b907cd12821f2e/pkg/sql/drop_schema.go#L185)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This needs to go to `dropSchemaImpl` otherwise the comment is not dropped upon DROP DATABASE CASCADE
</blockquote></details>

Fixed.
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5085 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqkgGH-Rwkp3gtO3wW:-MhKBD0OAgVka_aSHyeu:bis8y3h) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5085)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

Is there a test for it? We should keep it for sure if there is and add it if it isn't. 
</blockquote></details>

Yes, there's. Please, see the latest version of the PR.

https://github.com/cockroachdb/cockroach/pull/68606/files#diff-44361d449f2d21bb6beb453a14b8475fe98e73a04bf990d251ed99f055001eb8R5080
___
*[pkg/sql/logictest/testdata/logic_test/pg_catalog, line 5082 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MhGvFV98nv8A46mYcnL:-MhKBMYSCxBkbPpeGuoF:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b4e6fd8283f321fc119346558ad4644c6e7c14/pkg/sql/logictest/testdata/logic_test/pg_catalog#L5082)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: drop the trailing `;` here and above
</blockquote></details>

Fixed.
___
*[pkg/sql/logictest/testdata/logic_test/schema, line 881 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MhGvI9B97OtFUBj-9oI:-MhKBdy59HbItxUPbdH4:bt2vr1b) ([raw file](https://github.com/cockroachdb/cockroach/blob/84b4e6fd8283f321fc119346558ad4644c6e7c14/pkg/sql/logictest/testdata/logic_test/schema#L881)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

nit: drop the trailing `;` here 
</blockquote></details>

Fixed.

**Review by ekalinin — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/68606)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ekalinin, @knz, @otan, and @rafiss)

___
*[pkg/sql/sem/builtins/pg_builtins.go, line 1946 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/68606#-MgqaSox7ndfLf79FuDJ-r2-1946:-MhKDeBE3LtzfOTA-3OI:b-886gsx) ([raw file](https://github.com/cockroachdb/cockroach/blob/c02b84904998aadc09b690ad7a950664c3f5892e/pkg/sql/sem/builtins/pg_builtins.go#L1946)):*
<details><summary><i>Previously, otan (Oliver Tan) wrote…</i></summary><blockquote>

`pg_description` is correct
</blockquote></details>

Fixed

**Review by otan — APPROVED** [review]
looks good to me! thanks

bors r+

**Pull request merged**